## Credit Card Fraud Detection Using Machine Learning in Python

## About Dataset
## Context
It is important that credit card companies are able to recognize fraudulent credit card transactions so that customers are not charged for items that they did not purchase.

## Content
The dataset contains transactions made by credit cards in September 2013 by European cardholders.
This dataset presents transactions that occurred in two days, where we have 492 frauds out of 284,807 transactions. The dataset is highly unbalanced, the positive class (frauds) account for 0.172% of all transactions.

It contains only numerical input variables which are the result of a PCA transformation. Unfortunately, due to confidentiality issues, we cannot provide the original features and more background information about the data. Features V1, V2, … V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-sensitive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

Given the class imbalance ratio, we recommend measuring the accuracy using the Area Under the Precision-Recall Curve (AUPRC). Confusion matrix accuracy is not meaningful for unbalanced classification.

Importing the neccesary libraries

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [3]:
#loading the dataset
df=pd.read_csv("C:/Users/Shubham&Aryan/OneDrive/Documents/creditcard.csv")
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [4]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [12]:
#Checking the information of thedataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [8]:
#Checking the missing values
df.isnull().sum().sum()

0

* There are no missing values present in the dataset.

In [10]:
# Checking the distributon of legit transactions and fraudelent transactions
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

* There are 284315 legit transactions and 492 fraudulent transactions.
* The dataset is highly unbalanced.

0---> Normal transactions

1---> Fraud transactions

In [13]:
# seperating the data for analysis
legit= df[df.Class==0]
fraud= df[df.Class==1]

In [15]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [ ]:
# Statistical measures of the datset
legit.Amount.describe()

In [18]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [20]:
# compare the values for both transactions
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

##  Under-Sampling

* Build a sample dataset containing similar distribution of normal transactions     and Fraudulent Transactions
* Number of Fraudulent Transactions --> 492

In [22]:
legit_sample=legit.sample(n=492)

In [25]:
#Concatenating two dataframes
new_dataset=pd.concat([legit_sample,fraud],axis=0)
new_dataset

Time         V1        V2         V3        V4        V5  \
75819    56269.0   0.087050 -1.547541  -0.839388  1.600529 -0.058722   
175184  122232.0  -0.935674  0.129676   0.808992 -2.272691 -0.526710   
21514    31718.0 -12.918205  6.862129 -12.216383  1.391043 -5.445491   
41468    40690.0   1.108779 -0.291688   0.252184  0.588366 -0.242700   
104352   69030.0   1.279328 -1.439041   0.889018 -1.204807 -2.005251   
...          ...        ...       ...        ...       ...       ...   
279863  169142.0  -1.927883  1.125653  -4.518331  1.749293 -1.566487   
280143  169347.0   1.378559  1.289381  -5.004247  1.411850  0.442581   
280149  169351.0  -0.676143  1.126366  -2.213700  0.468308 -1.120541   
281144  169966.0  -3.113832  0.585864  -5.399730  1.817092 -0.840618   
281674  170348.0   1.991976  0.158476  -2.583441  0.408670  1.151147   

              V6        V7        V8        V9  ...       V21       V22  \
75819   0.331889  1.110655 -0.275008 -0.157169  ...  0.318999 -0.192974   
175184 -0.127746 -0.092546  0.368198 -0.999818  ...  0.440957  1.091978   
21514  -1.588276 -3.856525  6.599689  2.831173  ... -0.760840 -1.474006   
41468   0.226884 -0.173825  0.105050  0.681836  ... -0.290836 -0.812319   
104352 -0.465490 -1.329678  0.012025 -1.467877  ...  0.083640  0.502211   
...          ...       ...       ...       ...  ...       ...       ...   
279863 -2.010494 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189   
280143 -1.326536 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234   
280149 -0.003346 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108   
281144 -2.943548 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209   
281674 -0.096695  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  \
75819  -0.770419 -0.684518  0.699158 -0.247745 -0.081381  0.109981  608.85   
175184 -0.133063  0.657881  0.064046 -0.216655 -0.105290  0.051636   69.00   
21514  -0.855451 -1.687605  0.783973 -0.421366 -0.816828 -0.914960    4.49   
41468  -0.027753 -0.631412  0.292505  0.355894 -0.024396  0.014767   71.95   
104352 -0.021952  0.387862  0.203881 -0.043211  0.054059  0.039275   89.00   
...          ...       ...       ...       ...       ...       ...     ...   
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00   
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76   
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89   
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00   
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53   

        Class  
75819       0  
175184      0  
21514       0  
41468       0  
104352      0  
...       ...  
279863      1  
280143      1  
280149      1  
281144      1  
281674      1  

[984 rows x 31 columns]

In [27]:
new_dataset.shape

(984, 31)

In [29]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94661.493902 -0.071479  0.073291 -0.002519 -0.038818 -0.106756   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.024855  0.104108  0.030270  0.067954  ... -0.023910 -0.064675   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.067158 -0.012637 -0.025008  0.010129 -0.039573  0.004045 -0.026347   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       90.911443  
1      122.211321  

[2 rows x 30 columns]

In [31]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

## Splitting the data into Features & Targets

In [32]:
X=new_dataset.drop(columns='Class',axis=1)
Y=new_dataset['Class']

In [33]:
print(X)

            Time         V1        V2         V3        V4        V5  \
75819    56269.0   0.087050 -1.547541  -0.839388  1.600529 -0.058722   
175184  122232.0  -0.935674  0.129676   0.808992 -2.272691 -0.526710   
21514    31718.0 -12.918205  6.862129 -12.216383  1.391043 -5.445491   
41468    40690.0   1.108779 -0.291688   0.252184  0.588366 -0.242700   
104352   69030.0   1.279328 -1.439041   0.889018 -1.204807 -2.005251   
...          ...        ...       ...        ...       ...       ...   
279863  169142.0  -1.927883  1.125653  -4.518331  1.749293 -1.566487   
280143  169347.0   1.378559  1.289381  -5.004247  1.411850  0.442581   
280149  169351.0  -0.676143  1.126366  -2.213700  0.468308 -1.120541   
281144  169966.0  -3.113832  0.585864  -5.399730  1.817092 -0.840618   
281674  170348.0   1.991976  0.158476  -2.583441  0.408670  1.151147   

              V6        V7        V8        V9  ...       V20       V21  \
75819   0.331889  1.110655 -0.275008 -0.157169  ...  1.14255

In [34]:
print(Y)

75819     0
175184    0
21514     0
41468     0
104352    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


## Splitting the datset into Training data and Testing Data

In [35]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=2)

In [36]:
print(X.shape,X_train.shape,X_test.shape)

(984, 30) (787, 30) (197, 30)


## Model Training

## Logistic Regression

In [48]:
model=LogisticRegression()

In [49]:
# training the logistic Regression Model with Training Data
model.fit(X_train,Y_train)

C:\Users\Public\New\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

## Model Evaluation

## Accuracy Score

In [42]:
#accuracy on training data
X_train_prediction=model.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction,Y_train)

In [43]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9504447268106735


In [44]:
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [45]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.934010152284264
